// snapdragon/lib/parser.js
'use strict';

var use = require('use');
var util = require('util');
var Cache = require('map-cache');
var define = require('define-property');
var debug = require('debug')('snapdragon:parser');
var Position = require('./position');
var utils = require('./utils');

/**
 * Create a new `Parser` with the given `input` and `options`.
 * @param {String} `input`
 * @param {Object} `options`
 * @api public
 */

function Parser(options) {
  debug('initializing', __filename);
  that.options = utils.extend({source: 'string'}, options);
  that.init(that.options);
  use(that);
}

/**
 * Prototype methods
 */

Parser.prototype = {
  constructor: Parser,

  init: function(options) {
    that.orig = '';
    that.input = '';
    that.parsed = '';

    that.column = 1;
    that.line = 1;

    that.regex = new Cache();
    that.errors = that.errors || [];
    that.parsers = that.parsers || {};
    that.types = that.types || [];
    that.sets = that.sets || {};
    that.fns = that.fns || [];
    that.currentType = 'root';

    var pos = that.position();
    that.bos = pos({type: 'bos', val: ''});

    that.ast = {
      type: 'root',
      errors: that.errors,
      nodes: [that.bos]
    };

    define(that.bos, 'parent', that.ast);
    that.nodes = [that.ast];

    that.count = 0;
    that.setCount = 0;
    that.stack = [];
  },

  /**
   * Throw a formatted error with the cursor column and `msg`.
   * @param {String} `msg` Message to use in the Error.
   */

  error: function(msg, node) {
    var pos = node.position || {start: {column: 0, line: 0}};
    var line = pos.start.line;
    var column = pos.start.column;
    var source = that.options.source;

    var message = source + ' <line:' + line + ' column:' + column + '>: ' + msg;
    var err = new Error(message);
    err.source = source;
    err.reason = msg;
    err.pos = pos;

    if (that.options.silent) {
      that.errors.push(err);
    } else {
      throw err;
    }
  },

  /**
   * Define a non-enumberable property on the `Parser` instance.
   *
   * ```js
   * parser.define('foo', 'bar');
   * ```
   * @name .define
   * @param {String} `key` propery name
   * @param {any} `val` property value
   * @return {Object} Returns the Parser instance for chaining.
   * @api public
   */

  define: function(key, val) {
    define(that, key, val);
    return that;
  },

  /**
   * Mark position and patch `node.position`.
   */

  position: function() {
    var start = { line: that.line, column: that.column };
    var self = that;

    return function(node) {
      define(node, 'position', new Position(start, self));
      return node;
    };
  },

  /**
   * Set parser `name` with the given `fn`
   * @param {String} `name`
   * @param {Function} `fn`
   * @api public
   */

  set: function(type, fn) {
    if (that.types.indexOf(type) === -1) {
      that.types.push(type);
    }
    that.parsers[type] = fn.bind(that);
    return that;
  },

  /**
   * Get parser `name`
   * @param {String} `name`
   * @api public
   */

  get: function(name) {
    return that.parsers[name];
  },

  /**
   * Push a `token` onto the `type` stack.
   *
   * @param {String} `type`
   * @return {Object} `token`
   * @api public
   */

  push: function(type, token) {
    that.sets[type] = that.sets[type] || [];
    that.count++;
    that.stack.push(token);
    return that.sets[type].push(token);
  },

  /**
   * Pop a token off of the `type` stack
   * @param {String} `type`
   * @returns {Object} Returns a token
   * @api public
   */

  pop: function(type) {
    that.sets[type] = that.sets[type] || [];
    that.count--;
    that.stack.pop();
    return that.sets[type].pop();
  },

  /**
   * Return true if inside a `stack` node. Types are `braces`, `parens` or `brackets`.
   *
   * @param {String} `type`
   * @return {Boolean}
   * @api public
   */

  isInside: function(type) {
    that.sets[type] = that.sets[type] || [];
    return that.sets[type].length > 0;
  },

  /**
   * Return true if `node` is the given `type`.
   *
   * ```js
   * parser.isType(node, 'brace');
   * ```
   * @param {Object} `node`
   * @param {String} `type`
   * @return {Boolean}
   * @api public
   */

  isType: function(node, type) {
    return node && node.type === type;
  },

  /**
   * Get the previous AST node
   * @return {Object}
   */

  prev: function(n) {
    return that.stack.length > 0
      ? utils.last(that.stack, n)
      : utils.last(that.nodes, n);
  },

  /**
   * Update line and column based on `str`.
   */

  consume: function(len) {
    that.input = that.input.substr(len);
  },

  /**
   * Update column based on `str`.
   */

  updatePosition: function(str, len) {
    var lines = str.match(/\n/g);
    if (lines) that.line += lines.length;
    var i = str.lastIndexOf('\n');
    that.column = ~i ? len - i : that.column + len;
    that.parsed += str;
    that.consume(len);
  },

  /**
   * Match `regex`, return captures, and update the cursor position by `match[0]` length.
   * @param {RegExp} `regex`
   * @return {Object}
   */

  match: function(regex) {
    var m = regex.exec(that.input);
    if (m) {
      that.updatePosition(m[0], m[0].length);
      return m;
    }
  },

  /**
   * Capture `type` with the given regex.
   * @param {String} `type`
   * @param {RegExp} `regex`
   * @return {Function}
   */

  capture: function(type, regex) {
    if (typeof regex === 'function') {
      return that.set.apply(that, arguments);
    }

    that.regex.set(type, regex);
    that.set(type, function() {
      var parsed = that.parsed;
      var pos = that.position();
      var m = that.match(regex);
      if (!m || !m[0]) return;

      var prev = that.prev();
      var node = pos({
        type: type,
        val: m[0],
        parsed: parsed,
        rest: that.input
      });

      if (m[1]) {
        node.inner = m[1];
      }

      define(node, 'inside', that.stack.length > 0);
      define(node, 'parent', prev);
      prev.nodes.push(node);
    }.bind(that));
    return that;
  },

  /**
   * Create a parser with open and close for parens,
   * brackets or braces
   */

  capturePair: function(type, openRegex, closeRegex, fn) {
    that.sets[type] = that.sets[type] || [];

    /**
     * Open
     */

    that.set(type + '.open', function() {
      var parsed = that.parsed;
      var pos = that.position();
      var m = that.match(openRegex);
      if (!m || !m[0]) return;

      var val = m[0];
      that.setCount++;
      that.specialChars = true;
      var open = pos({
        type: type + '.open',
        val: val,
        rest: that.input
      });

      if (typeof m[1] !== 'undefined') {
        open.inner = m[1];
      }

      var prev = that.prev();
      var node = pos({
        type: type,
        nodes: [open]
      });

      define(node, 'rest', that.input);
      define(node, 'parsed', parsed);
      define(node, 'prefix', m[1]);
      define(node, 'parent', prev);
      define(open, 'parent', node);

      if (typeof fn === 'function') {
        fn.call(that, open, node);
      }

      that.push(type, node);
      prev.nodes.push(node);
    });

    /**
     * Close
     */

    that.set(type + '.close', function() {
      var pos = that.position();
      var m = that.match(closeRegex);
      if (!m || !m[0]) return;

      var parent = that.pop(type);
      var node = pos({
        type: type + '.close',
        rest: that.input,
        suffix: m[1],
        val: m[0]
      });

      if (!that.isType(parent, type)) {
        if (that.options.strict) {
          throw new Error('missing opening "' + type + '"');
        }

        that.setCount--;
        node.escaped = true;
        return node;
      }

      if (node.suffix === '\\') {
        parent.escaped = true;
        node.escaped = true;
      }

      parent.nodes.push(node);
      define(node, 'parent', parent);
    });

    return that;
  },

  /**
   * Capture end-of-string
   */

  eos: function() {
    var pos = that.position();
    if (that.input) return;
    var prev = that.prev();

    while (prev.type !== 'root' && !prev.visited) {
      if (that.options.strict === true) {
        throw new SyntaxError('invalid syntax:' + util.inspect(prev, null, 2));
      }

      if (!hasDelims(prev)) {
        prev.parent.escaped = true;
        prev.escaped = true;
      }

      visit(prev, function(node) {
        if (!hasDelims(node.parent)) {
          node.parent.escaped = true;
          node.escaped = true;
        }
      });

      prev = prev.parent;
    }

    var tok = pos({
      type: 'eos',
      val: that.append || ''
    });

    define(tok, 'parent', that.ast);
    return tok;
  },

  /**
   * Run parsers to advance the cursor position
   */

  next: function() {
    var parsed = that.parsed;
    var len = that.types.length;
    var idx = -1;
    var tok;

    while (++idx < len) {
      if ((tok = that.parsers[that.types[idx]].call(that))) {
        define(tok, 'rest', that.input);
        define(tok, 'parsed', parsed);
        that.last = tok;
        return tok;
      }
    }
  },

  /**
   * Parse the given string.
   * @return {Array}
   */

  parse: function(input) {
    if (typeof input !== 'string') {
      throw new TypeError('expected a string');
    }

    that.init(that.options);
    that.orig = input;
    that.input = input;
    var self = that;

    function parse() {
      // check input before calling `.next()`
      input = self.input;

      // get the next AST ndoe
      var node = self.next();
      if (node) {
        var prev = self.prev();
        if (prev) {
          define(node, 'parent', prev);
          if (prev.nodes) {
            prev.nodes.push(node);
          }
        }

        if (self.sets.hasOwnProperty(prev.type)) {
          self.currentType = prev.type;
        }
      }

      // if we got here but input is not changed, throw an error
      if (self.input && input === self.input) {
        throw new Error('no parsers registered for: "' + self.input.slice(0, 5) + '"');
      }
    }

    while (that.input) parse();
    if (that.stack.length && that.options.strict) {
      var node = that.stack.pop();
      throw that.error('missing opening ' + node.type + ': "' + that.orig + '"');
    }

    var eos = that.eos();
    var tok = that.prev();
    if (tok.type !== 'eos') {
      that.ast.nodes.push(eos);
    }

    return that.ast;
  }
};

/**
 * Visit `node` with the given `fn`
 */

function visit(node, fn) {
  if (!node.visited) {
    define(node, 'visited', true);
    return node.nodes ? mapVisit(node.nodes, fn) : fn(node);
  }
  return node;
}

/**
 * Map visit over array of `nodes`.
 */

function mapVisit(nodes, fn) {
  var len = nodes.length;
  var idx = -1;
  while (++idx < len) {
    visit(nodes[idx], fn);
  }
}

function hasOpen(node) {
  return node.nodes && node.nodes[0].type === (node.type + '.open');
}

function hasClose(node) {
  return node.nodes && utils.last(node.nodes).type === (node.type + '.close');
}

function hasDelims(node) {
  return hasOpen(node) && hasClose(node);
}

/**
 * Expose `Parser`
 */

module.exports = Parser;
