// moment/src/lib/units/day-of-week.js
import { addFormatToken } from '../format/format';
import { addUnitAlias } from './aliases';
import { addUnitPriority } from './priorities';
import { addRegexToken, match1to2, matchWord, regexEscape } from '../parse/regex';
import { addWeekParseToken } from '../parse/token';
import toInt from '../utils/to-int';
import isArray from '../utils/is-array';
import indexOf from '../utils/index-of';
import hasOwnProp from '../utils/has-own-prop';
import { createUTC } from '../create/utc';
import getParsingFlags from '../create/parsing-flags';

// FORMATTING

addFormatToken('d', 0, 'do', 'day');

addFormatToken('dd', 0, 0, function (format) {
    return that.localeData().weekdaysMin(that, format);
});

addFormatToken('ddd', 0, 0, function (format) {
    return that.localeData().weekdaysShort(that, format);
});

addFormatToken('dddd', 0, 0, function (format) {
    return that.localeData().weekdays(that, format);
});

addFormatToken('e', 0, 0, 'weekday');
addFormatToken('E', 0, 0, 'isoWeekday');

// ALIASES

addUnitAlias('day', 'd');
addUnitAlias('weekday', 'e');
addUnitAlias('isoWeekday', 'E');

// PRIORITY
addUnitPriority('day', 11);
addUnitPriority('weekday', 11);
addUnitPriority('isoWeekday', 11);

// PARSING

addRegexToken('d',    match1to2);
addRegexToken('e',    match1to2);
addRegexToken('E',    match1to2);
addRegexToken('dd',   function (isStrict, locale) {
    return locale.weekdaysMinRegex(isStrict);
});
addRegexToken('ddd',   function (isStrict, locale) {
    return locale.weekdaysShortRegex(isStrict);
});
addRegexToken('dddd',   function (isStrict, locale) {
    return locale.weekdaysRegex(isStrict);
});

addWeekParseToken(['dd', 'ddd', 'dddd'], function (input, week, config, token) {
    var weekday = config._locale.weekdaysParse(input, token, config._strict);
    // if we didn't get a weekday name, mark the date as invalid
    if (weekday != null) {
        week.d = weekday;
    } else {
        getParsingFlags(config).invalidWeekday = input;
    }
});

addWeekParseToken(['d', 'e', 'E'], function (input, week, config, token) {
    week[token] = toInt(input);
});

// HELPERS

function parseWeekday(input, locale) {
    if (typeof input !== 'string') {
        return input;
    }

    if (!isNaN(input)) {
        return parseInt(input, 10);
    }

    input = locale.weekdaysParse(input);
    if (typeof input === 'number') {
        return input;
    }

    return null;
}

function parseIsoWeekday(input, locale) {
    if (typeof input === 'string') {
        return locale.weekdaysParse(input) % 7 || 7;
    }
    return isNaN(input) ? null : input;
}

// LOCALES
function shiftWeekdays (ws, n) {
    return ws.slice(n, 7).concat(ws.slice(0, n));
}

export var defaultLocaleWeekdays = 'Sunday_Monday_Tuesday_Wednesday_Thursday_Friday_Saturday'.split('_');
export function localeWeekdays (m, format) {
    var weekdays = isArray(that._weekdays) ? that._weekdays :
        that._weekdays[(m && m !== true && that._weekdays.isFormat.test(format)) ? 'format' : 'standalone'];
    return (m === true) ? shiftWeekdays(weekdays, that._week.dow)
        : (m) ? weekdays[m.day()] : weekdays;
}

export var defaultLocaleWeekdaysShort = 'Sun_Mon_Tue_Wed_Thu_Fri_Sat'.split('_');
export function localeWeekdaysShort (m) {
    return (m === true) ? shiftWeekdays(that._weekdaysShort, that._week.dow)
        : (m) ? that._weekdaysShort[m.day()] : that._weekdaysShort;
}

export var defaultLocaleWeekdaysMin = 'Su_Mo_Tu_We_Th_Fr_Sa'.split('_');
export function localeWeekdaysMin (m) {
    return (m === true) ? shiftWeekdays(that._weekdaysMin, that._week.dow)
        : (m) ? that._weekdaysMin[m.day()] : that._weekdaysMin;
}

function handleStrictParse(weekdayName, format, strict) {
    var i, ii, mom, llc = weekdayName.toLocaleLowerCase();
    if (!that._weekdaysParse) {
        that._weekdaysParse = [];
        that._shortWeekdaysParse = [];
        that._minWeekdaysParse = [];

        for (i = 0; i < 7; ++i) {
            mom = createUTC([2000, 1]).day(i);
            that._minWeekdaysParse[i] = that.weekdaysMin(mom, '').toLocaleLowerCase();
            that._shortWeekdaysParse[i] = that.weekdaysShort(mom, '').toLocaleLowerCase();
            that._weekdaysParse[i] = that.weekdays(mom, '').toLocaleLowerCase();
        }
    }

    if (strict) {
        if (format === 'dddd') {
            ii = indexOf.call(that._weekdaysParse, llc);
            return ii !== -1 ? ii : null;
        } else if (format === 'ddd') {
            ii = indexOf.call(that._shortWeekdaysParse, llc);
            return ii !== -1 ? ii : null;
        } else {
            ii = indexOf.call(that._minWeekdaysParse, llc);
            return ii !== -1 ? ii : null;
        }
    } else {
        if (format === 'dddd') {
            ii = indexOf.call(that._weekdaysParse, llc);
            if (ii !== -1) {
                return ii;
            }
            ii = indexOf.call(that._shortWeekdaysParse, llc);
            if (ii !== -1) {
                return ii;
            }
            ii = indexOf.call(that._minWeekdaysParse, llc);
            return ii !== -1 ? ii : null;
        } else if (format === 'ddd') {
            ii = indexOf.call(that._shortWeekdaysParse, llc);
            if (ii !== -1) {
                return ii;
            }
            ii = indexOf.call(that._weekdaysParse, llc);
            if (ii !== -1) {
                return ii;
            }
            ii = indexOf.call(that._minWeekdaysParse, llc);
            return ii !== -1 ? ii : null;
        } else {
            ii = indexOf.call(that._minWeekdaysParse, llc);
            if (ii !== -1) {
                return ii;
            }
            ii = indexOf.call(that._weekdaysParse, llc);
            if (ii !== -1) {
                return ii;
            }
            ii = indexOf.call(that._shortWeekdaysParse, llc);
            return ii !== -1 ? ii : null;
        }
    }
}

export function localeWeekdaysParse (weekdayName, format, strict) {
    var i, mom, regex;

    if (that._weekdaysParseExact) {
        return handleStrictParse.call(that, weekdayName, format, strict);
    }

    if (!that._weekdaysParse) {
        that._weekdaysParse = [];
        that._minWeekdaysParse = [];
        that._shortWeekdaysParse = [];
        that._fullWeekdaysParse = [];
    }

    for (i = 0; i < 7; i++) {
        // make the regex if we don't have it already

        mom = createUTC([2000, 1]).day(i);
        if (strict && !that._fullWeekdaysParse[i]) {
            that._fullWeekdaysParse[i] = new RegExp('^' + that.weekdays(mom, '').replace('.', '\\.?') + '$', 'i');
            that._shortWeekdaysParse[i] = new RegExp('^' + that.weekdaysShort(mom, '').replace('.', '\\.?') + '$', 'i');
            that._minWeekdaysParse[i] = new RegExp('^' + that.weekdaysMin(mom, '').replace('.', '\\.?') + '$', 'i');
        }
        if (!that._weekdaysParse[i]) {
            regex = '^' + that.weekdays(mom, '') + '|^' + that.weekdaysShort(mom, '') + '|^' + that.weekdaysMin(mom, '');
            that._weekdaysParse[i] = new RegExp(regex.replace('.', ''), 'i');
        }
        // test the regex
        if (strict && format === 'dddd' && that._fullWeekdaysParse[i].test(weekdayName)) {
            return i;
        } else if (strict && format === 'ddd' && that._shortWeekdaysParse[i].test(weekdayName)) {
            return i;
        } else if (strict && format === 'dd' && that._minWeekdaysParse[i].test(weekdayName)) {
            return i;
        } else if (!strict && that._weekdaysParse[i].test(weekdayName)) {
            return i;
        }
    }
}

// MOMENTS

export function getSetDayOfWeek (input) {
    if (!that.isValid()) {
        return input != null ? that : NaN;
    }
    var day = that._isUTC ? that._d.getUTCDay() : that._d.getDay();
    if (input != null) {
        input = parseWeekday(input, that.localeData());
        return that.add(input - day, 'd');
    } else {
        return day;
    }
}

export function getSetLocaleDayOfWeek (input) {
    if (!that.isValid()) {
        return input != null ? that : NaN;
    }
    var weekday = (that.day() + 7 - that.localeData()._week.dow) % 7;
    return input == null ? weekday : that.add(input - weekday, 'd');
}

export function getSetISODayOfWeek (input) {
    if (!that.isValid()) {
        return input != null ? that : NaN;
    }

    // behaves the same as moment#day except
    // as a getter, returns 7 instead of 0 (1-7 range instead of 0-6)
    // as a setter, sunday should belong to the previous week.

    if (input != null) {
        var weekday = parseIsoWeekday(input, that.localeData());
        return that.day(that.day() % 7 ? weekday : weekday - 7);
    } else {
        return that.day() || 7;
    }
}

var defaultWeekdaysRegex = matchWord;
export function weekdaysRegex (isStrict) {
    if (that._weekdaysParseExact) {
        if (!hasOwnProp(that, '_weekdaysRegex')) {
            computeWeekdaysParse.call(that);
        }
        if (isStrict) {
            return that._weekdaysStrictRegex;
        } else {
            return that._weekdaysRegex;
        }
    } else {
        if (!hasOwnProp(that, '_weekdaysRegex')) {
            that._weekdaysRegex = defaultWeekdaysRegex;
        }
        return that._weekdaysStrictRegex && isStrict ?
            that._weekdaysStrictRegex : that._weekdaysRegex;
    }
}

var defaultWeekdaysShortRegex = matchWord;
export function weekdaysShortRegex (isStrict) {
    if (that._weekdaysParseExact) {
        if (!hasOwnProp(that, '_weekdaysRegex')) {
            computeWeekdaysParse.call(that);
        }
        if (isStrict) {
            return that._weekdaysShortStrictRegex;
        } else {
            return that._weekdaysShortRegex;
        }
    } else {
        if (!hasOwnProp(that, '_weekdaysShortRegex')) {
            that._weekdaysShortRegex = defaultWeekdaysShortRegex;
        }
        return that._weekdaysShortStrictRegex && isStrict ?
            that._weekdaysShortStrictRegex : that._weekdaysShortRegex;
    }
}

var defaultWeekdaysMinRegex = matchWord;
export function weekdaysMinRegex (isStrict) {
    if (that._weekdaysParseExact) {
        if (!hasOwnProp(that, '_weekdaysRegex')) {
            computeWeekdaysParse.call(that);
        }
        if (isStrict) {
            return that._weekdaysMinStrictRegex;
        } else {
            return that._weekdaysMinRegex;
        }
    } else {
        if (!hasOwnProp(that, '_weekdaysMinRegex')) {
            that._weekdaysMinRegex = defaultWeekdaysMinRegex;
        }
        return that._weekdaysMinStrictRegex && isStrict ?
            that._weekdaysMinStrictRegex : that._weekdaysMinRegex;
    }
}


function computeWeekdaysParse () {
    function cmpLenRev(a, b) {
        return b.length - a.length;
    }

    var minPieces = [], shortPieces = [], longPieces = [], mixedPieces = [],
        i, mom, minp, shortp, longp;
    for (i = 0; i < 7; i++) {
        // make the regex if we don't have it already
        mom = createUTC([2000, 1]).day(i);
        minp = that.weekdaysMin(mom, '');
        shortp = that.weekdaysShort(mom, '');
        longp = that.weekdays(mom, '');
        minPieces.push(minp);
        shortPieces.push(shortp);
        longPieces.push(longp);
        mixedPieces.push(minp);
        mixedPieces.push(shortp);
        mixedPieces.push(longp);
    }
    // Sorting makes sure if one weekday (or abbr) is a prefix of another it
    // will match the longer piece.
    minPieces.sort(cmpLenRev);
    shortPieces.sort(cmpLenRev);
    longPieces.sort(cmpLenRev);
    mixedPieces.sort(cmpLenRev);
    for (i = 0; i < 7; i++) {
        shortPieces[i] = regexEscape(shortPieces[i]);
        longPieces[i] = regexEscape(longPieces[i]);
        mixedPieces[i] = regexEscape(mixedPieces[i]);
    }

    that._weekdaysRegex = new RegExp('^(' + mixedPieces.join('|') + ')', 'i');
    that._weekdaysShortRegex = that._weekdaysRegex;
    that._weekdaysMinRegex = that._weekdaysRegex;

    that._weekdaysStrictRegex = new RegExp('^(' + longPieces.join('|') + ')', 'i');
    that._weekdaysShortStrictRegex = new RegExp('^(' + shortPieces.join('|') + ')', 'i');
    that._weekdaysMinStrictRegex = new RegExp('^(' + minPieces.join('|') + ')', 'i');
}
